java/ql/src/semmle/code/java/security/UnsafeDeserialization.qll

```diff
@@ -114,6 +121,116 @@ class SafeKryo extends DataFlow2::Configuration {
   }
 }
 
+/**
+ * A method that overrides `android.os.Parcelable.Creator.createFromParcel`.
+ */
+class CreateFromParcelMethod extends Method {
+  CreateFromParcelMethod() {
+    this.hasName("createFromParcel") and
+    this.getEnclosingCallable().getDeclaringType().getASupertype*() instanceof TypeParcelable
+  }
+}
+
+/** Unsafe Gson configuration with dynamic type. */
+class UnsafeGsonConfig extends TaintTracking2::Configuration {
+  UnsafeGsonConfig() { this = "UnsafeDeserialization::UnsafeGsonConfig" }
+
+  override predicate isSource(DataFlow::Node src) { src instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof GsonDeserializeMethod and
+      sink.asExpr() = ma.getArgument(1) // The class type argument
+    )
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(ReflectiveClassIdentifierMethodAccess ma |
+      node1.asExpr() = ma.getArgument(0) and
+      node2.asExpr() = ma
+    )
```

Comment:
Lift this into `private predicate isReflectiveClassIdentifierStep(DataFlow::Node node1, DataFlow::Node node2)` and use it at the three callsites

Reply:
Changed